src/pudl/analysis/plant_parts_eia.py

```diff
@@ -0,0 +1,1530 @@
+
+`PLANT_PARTS` is the main recipe book for how each of the plant-parts need to
+be compiled.
+
+The three main classes which enable the generation of the plant-part table are:
+* ``MakeMegaGenTbl``: All of the plant parts are compiled from generators. So
+this class generates a big dataframe of generators with any ID and data columns
+we'll need. This is also where we add records regarding utility ownership
+slices. The table includes two records for every generator-owner: one for the
+"total" generator (assuming the owner owns 100% of the generator) and one for
+the report ownership fraction of that generator with all of the data columns
+scaled to the ownership fraction.
+* ``LabelTrueGranularities``: This class creates labels for all generators
+which note wether the plant-part records that will be compiled from each
+generator will be a "true granulary", as described above.
+* ``MakePlantParts``: This class uses the generator dataframe, the granularity
+dataframe from the above two classes as well as the information stored in
+`PLANT_PARTS` to know how to aggregate each of the plant parts. Then we have
+plant part dataframes with the columns which identify the plant part and all of
+the data columns aggregated to the level of the plant part. With that compiled
+plant part dataframe we also add in qualifier columns with `AddQualifier`. A
+qualifer column is a column which contain data that is not endemic to the plant
+part record (it is not one of the identifying columns or aggregated data
+columns) but the data is still useful data that is attributable to each of the
+plant part records. For more detail on what a qualifier column is, see the
+`AddQualifier.execute()` method.
+
+Lines needed to generate the plant-parts df:
+``
+import pudl
+# make the pudl_out object
+pudl_engine = sa.create_engine(pudl.workspace.setup.get_defaults()['pudl_db'])
+pudl_out = pudl.output.pudltabl.PudlTabl(pudl_engine,freq='AS')
+``
+AND:
+Make the table via pudl_out:
+``
+plant_parts_eia = pudl_out.plant_parts_eia()
+``
+
+OR
+Make the table via objects in this module:
+``
+mega_gens = MakeMegaGenTbl(pudl_out).execute()
+true_grans = LabelTrueGranularities(mega_gens).execute()
+
+parts_compiler = MakePlantParts(pudl_out, mega_gens=mega_gens, true_grans=true_grans)
+plant_parts_eia = parts_compiler.execute()
+``
+
+"""
+
+import logging
+import warnings
+from copy import deepcopy
+
+import numpy as np
+import pandas as pd
+
+import pudl
+
+logger = logging.getLogger(__name__)
+
+PLANT_PARTS = {
+    'plant': {
+        'id_cols': ['plant_id_eia'],
+    },
+    'plant_gen': {
+        'id_cols': ['plant_id_eia', 'generator_id'],
+    },
+    'plant_unit': {
+        'id_cols': ['plant_id_eia', 'unit_id_pudl'],
+    },
+    'plant_technology': {
+        'id_cols': ['plant_id_eia', 'technology_description'],
+    },
+    'plant_prime_fuel': {
+        'id_cols': ['plant_id_eia', 'energy_source_code_1'],
+    },
```

Comment:
this is a very good point! I'd lean towards using `plant_primary_fuel` because `plant_primary_energy_source` is tooo looong but just `plant_energy_source` is not as clear given that the plants can have many different energy source codes.

Reply:
just fyi, I am flagging this as a thing to change but not changing it right now. When I do this, I need to convert all of the saved training data because the name of the plant-part is incorporated into the record id (which is the thing that is saved in the training data). I'll make an issue for this and do it in one fell swoop.